File: javascript/src/color.js
/**
 * @fileoverview A wrapper for colors.
 */

goog.provide('ee.Color');

goog.require('ee.ApiFunction');
goog.require('ee.ComputedObject');
goog.require('ee.Types');



/**
 * An object to represent colors.
 *
 * @param {string|!Array<number>|!ee.ComputedObject} color
 *    1) A W3C compatible color string.
 *    2) A list of RGBA values in the range of [0:1].
 *    3) A ComputedObject returning a color.
 * @constructor
 * @extends {ee.ComputedObject}
 * @export
 */
ee.Color = function(color) {
  if (!(this instanceof ee.Color)) {
    return new ee.Color(color);
  }

  ee.Color.initialize();

  if (color instanceof ee.ComputedObject) {
    ee.Color.base(this, 'constructor', color.func, color.args, color.varName);
  } else {
    ee.Color.base(
        this, 'constructor', new ee.ApiFunction('Color'),
        {'input': color});
  }
};
goog.inherits(ee.Color, ee.ComputedObject);


/**
 * Whether the class has been initialized.
 * @type {boolean}
 * @private
 */
ee.Color.initialized_ = false;


/**
 * Imports API functions to this class.
 */
ee.Color.initialize = function() {
  if (!ee.Color.initialized_) {
    ee.ApiFunction.importApi(ee.Color, 'Color', 'Color');
    ee.Color.initialized_ = true;
  }
};


/**
 * Removes imported API functions from this class.
 */
ee.Color.reset = function() {
  ee.ApiFunction.clearApi(ee.Color);
  ee.Color.initialized_ = false;
};


/** @override */
ee.Color.prototype.name = function() {
  return 'Color';
};


/**
 * Creates a Color given a list of HSV values.
 *
 * @param {!Array<number>|!ee.List} hsv A list of HSV (hue, saturation, value)
 *     values in the range [0:1].
 * @return {!ee.Color}
 * @export
 */
ee.Color.fromHsv = function(hsv) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.fromHsv', hsv));
};


/**
 * Creates a Color given a list of HSL values.
 *
 * @param {!Array<number>|!ee.List} hsl A list of HSL (hue, saturation,
 *     luminosity) values in the range [0:1].
 * @return {!ee.Color}
 * @export
 */
ee.Color.fromHsl = function(hsl) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.fromHsl', hsl));
};


/**
 * Creates a Color given a list of CIE-LAB values.
 *
 * @param {!Array<number>|!ee.List} lab A list of CIE-LAB values.
 * @return {!ee.Color}
 * @export
 */
ee.Color.fromLab = function(lab) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.fromLab', lab));
};


/**
 * Creates a Color given a list of CIE-LCH values.
 *
 * @param {!Array<number>|!ee.List} lch A list of CIE-LCH (lightness, chroma,
 *     hue) values.
 * @return {!ee.Color}
 * @export
 */
ee.Color.fromLch = function(lch) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.fromLch', lch));
};


/**
 * Creates a gray color.
 *
 * @param {number|!ee.Number} value The gray value in the range [0:1].
 * @param {(number|!ee.Number)=} opt_alpha The alpha value in the range [0:1].
 * @return {!ee.Color}
 * @export
 */
ee.Color.gray = function(value, opt_alpha) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.gray', value, opt_alpha));
};


/**
 * Mixes two colors.
 *
 * @param {!ee.Color} start The starting color.
 * @param {!ee.Color} end The ending color.
 * @param {(number|!ee.Number)=} opt_ratio The mix ratio.
 * @param {(string|!ee.String)=} opt_colorspace The colorspace to mix in.
 * @return {!ee.Color}
 * @export
 */
ee.Color.mix = function(start, end, opt_ratio, opt_colorspace) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.mix', start, end, opt_ratio, opt_colorspace));
};


/**
 * Scale each of the RGB channels to produce a brighter color.
 *
 * @param {(number|!ee.Number)=} opt_scale The scale factor.
 * @return {!ee.Color}
 * @export
 */
ee.Color.prototype.brighter = function(opt_scale) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.brighter', this, opt_scale));
};


/**
 * Scale each of the RGB channels to produce a darker color.
 *
 * @param {(number|!ee.Number)=} opt_scale The scale factor.
 * @return {!ee.Color}
 * @export
 */
ee.Color.prototype.darker = function(opt_scale) {
  return /** @type {!ee.Color} */ (
      ee.ApiFunction._call('Color.darker', this, opt_scale));
};


/**
 * Convert a color to HSL.
 *
 * @return {!ee.List}
 * @export
 */
ee.Color.prototype.toHsl = function() {
  return /** @type {!ee.List} */ (
      ee.ApiFunction._call('Color.toHsl', this));
};


/**
 * Convert a color to HSV.
 *
 * @return {!ee.List}
 * @export
 */
ee.Color.prototype.toHsv = function() {
  return /** @type {!ee.List} */ (
      ee.ApiFunction._call('Color.toHsv', this));
};


/**
 * Convert a color to CIE-Lab.
 *
 * @return {!ee.List}
 * @export
 */
ee.Color.prototype.toLab = function() {
  return /** @type {!ee.List} */ (
      ee.ApiFunction._call('Color.toLab', this));
};


/**
 * Convert a color to CIE-LCH.
 *
 * @return {!ee.List}
 * @export
 */
ee.Color.prototype.toLch = function() {
  return /** @type {!ee.List} */ (
      ee.ApiFunction._call('Color.toLch', this));
};


/**
 * Convert a color to RGB.
 *
 * @return {!ee.List}
 * @export
 */
ee.Color.prototype.toRgb = function() {
  return /** @type {!ee.List} */ (
      ee.ApiFunction._call('Color.toRGB', this));
};


/**
 * Returns value of a color as an RGBA hex string.
 *
 * @return {!ee.String}
 * @export
 */
ee.Color.prototype.toHexString = function() {
  return /** @type {!ee.String} */ (
      ee.ApiFunction._call('Color.toHexString', this));
};

File: python/ee/color.py
"""A wrapper for colors."""

from __future__ import annotations

from ee import _arg_types
from ee import apifunction
from ee import computedobject
from ee import ee_exception
from ee import ee_list
from ee import ee_string


class Color(computedobject.ComputedObject):
  """A color."""

  _initialized = False

  # Tell pytype to not complain about dynamic attributes.
  _HAS_DYNAMIC_ATTRIBUTES = True

  def __init__(self, input: _arg_types.Color | None = None):  # pylint: disable=redefined-builtin
    """Initializes a color wrapper.

    This constructor accepts the following args:
      1) A W3C compatible color string.
      2) A list of RGBA values in the range of [0:1].
      3) A ComputedObject returning a color.

    Args:
      input: The color to wrap.
    """
    self.initialize()

    if input is None:
      message = f'Required argument (input) missing to function: {self.name()}'
      raise ee_exception.EEException(message)

    if isinstance(input, computedobject.ComputedObject):
      super().__init__(input.func, input.args, input.varName)
    else:
      super().__init__(apifunction.ApiFunction(self.name()), {'input': input})

  @classmethod
  def initialize(cls) -> None:
    """Imports API functions to this class."""
    if not cls._initialized:
      apifunction.ApiFunction.importApi(cls, cls.name(), cls.name())
      cls._initialized = True

  @classmethod
  def reset(cls) -> None:
    """Removes imported API functions from this class."""
    apifunction.ApiFunction.clearApi(cls)
    cls._initialized = False

  @staticmethod
  def name() -> str:
    return 'Color'

  @classmethod
  def fromHsv(cls, hsv: _arg_types.List) -> Color:
    """Creates a Color given a list of HSV values.

    Args:
      hsv: A list of HSV (hue, saturation, value) values in the range [0:1].

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(cls.name() + '.fromHsv', hsv)

  @classmethod
  def fromHsl(cls, hsl: _arg_types.List) -> Color:
    """Creates a Color given a list of HSL values.

    Args:
      hsl: A list of HSL (hue, saturation, luminosity) values in the range
        [0:1].

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(cls.name() + '.fromHsl', hsl)

  @classmethod
  def fromLab(cls, lab: _arg_types.List) -> Color:
    """Creates a Color given a list of CIE-LAB values.

    Args:
      lab: A list of CIE-LAB values.

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(cls.name() + '.fromLab', lab)

  @classmethod
  def fromLch(cls, lch: _arg_types.List) -> Color:
    """Creates a Color given a list of CIE-LCH values.

    Args:
      lch: A list of CIE-LCH (lightness, chroma, hue) values.

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(cls.name() + '.fromLch', lch)

  @classmethod
  def gray(
      cls, value: _arg_types.Number, alpha: _arg_types.Number = 1.0
  ) -> Color:
    """Creates a gray color.

    Args:
      value: The gray value in the range [0:1].
      alpha: The alpha value in the range [0:1].

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(cls.name() + '.gray', value, alpha)

  def brighter(self, scale: _arg_types.Number = 0.7) -> Color:
    """Scales each of the RGB channels to produce a brighter color.

    Args:
      scale: The scale factor.

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(self.name() + '.brighter', self, scale)

  def darker(self, scale: _arg_types.Number = 0.7) -> Color:
    """Scales each of the RGB channels to produce a darker color.

    Args:
      scale: The scale factor.

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(self.name() + '.darker', self, scale)

  def mix(
      self,
      end: _arg_types.Color,
      ratio: _arg_types.Number = 0.5,
      colorspace: _arg_types.String = 'rgb',
  ) -> Color:
    """Mixes two colors.

    Args:
      end: The ending color.
      ratio: The mix ratio.
      colorspace: The colorspace to mix in. One of 'rgb', 'lrgb', 'hsv', 'hsl',
        'lab', or 'lch'.

    Returns:
      An ee.Color.
    """
    return apifunction.ApiFunction.call_(
        self.name() + '.mix', self, end, ratio, colorspace
    )

  def toHexString(self) -> ee_string.String:
    """Returns value of a color as an RGBA hex string."""
    return apifunction.ApiFunction.call_(self.name() + '.toHexString', self)

  def toHsl(self) -> ee_list.List:
    """Returns Hue, Saturation, Lightness (HSL) as a List in the range [0:1]."""
    return apifunction.ApiFunction.call_(self.name() + '.toHsl', self)

  def toHsv(self) -> ee_list.List:
    """Returns Hue Saturation Value (HSV) as a list in the range [0:1]."""
    return apifunction.ApiFunction.call_(self.name() + '.toHsv', self)

  def toLab(self) -> ee_list.List:
    """Returns the CIE-Lab as a list of values.

    The components are unnormalized and have the approximate ranges of:

    L=[0:100], a,b=[-127:128].
    """
    return apifunction.ApiFunction.call_(self.name() + '.toLab', self)

  def toLch(self) -> ee_list.List:
    """Returns the CIE-LCH as a list of values.

    The components are unnormalized and have an approximate range of:

    L=[0:100], C=[0:180], H=[0:360] (degrees).
    """
    return apifunction.ApiFunction.call_(self.name() + '.toLch', self)

  def toRGB(self) -> ee_list.List:
    """Returns RGB and alpha as a list of numbers in the range of [0:1].
    """
    return apifunction.ApiFunction.call_(self.name() + '.toRGB', self)
